samplePDF/samplePDFFDBase.cpp

```diff
@@ -340,328 +400,289 @@ void samplePDFFDBase::fillArray() {
+	//similarly to XVar now
+	
+	//DB Find the relevant bin in the PDF for each event
+	int XBinToFill = -1;
+	int YBinToFill = MCSamples[iSample].NomYBin[iEvent];
+	
+	//DB Check to see if momentum shift has moved bins
+	//DB - First, check to see if the event is still in the nominal bin	
+	if (XVar < MCSamples[iSample].rw_upper_xbinedge[iEvent] && XVar >= MCSamples[iSample].rw_lower_xbinedge[iEvent]) {
+	  XBinToFill = MCSamples[iSample].NomXBin[iEvent];
+	}
+	//DB - Second, check to see if the event is outside of the binning range and skip event if it is
+	else if (XVar < XBinEdges[0] || XVar >= XBinEdges[nXBins]) {
+	  continue;
+	}
+	//DB - Thirdly, check the adjacent bins first as Eb+CC+EScale shifts aren't likely to move an Erec more than 1bin width
+	//Shifted down one bin from the event bin at nominal
+	else if (XVar < MCSamples[iSample].rw_lower_xbinedge[iEvent] && XVar >= MCSamples[iSample].rw_lower_lower_xbinedge[iEvent]) {
+	  XBinToFill = MCSamples[iSample].NomXBin[iEvent]-1;
+	}
+	//Shifted up one bin from the event bin at nominal
+	else if (XVar < MCSamples[iSample].rw_upper_upper_xbinedge[iEvent] && XVar >= MCSamples[iSample].rw_upper_xbinedge[iEvent]) {
+	  XBinToFill = MCSamples[iSample].NomXBin[iEvent]+1;
+	}
+	//DB - If we end up in this loop, the event has been shifted outside of its nominal bin, but is still within the allowed binning range
+	else {
+	  for(unsigned int iBin=0;iBin<(XBinEdges.size()-1);iBin++) {
+	    if (XVar >= XBinEdges[iBin] && XVar < XBinEdges[iBin+1]) {
+	      XBinToFill = iBin;
+	    }
+	  }
+	}
+	
+	//ETA - we can probably remove this final if check on the -1? 
+	//Maybe we can add an overflow bin to the array and assign any events to this bin?
+	//Might save us an extra if call?
+	//DB Fill relevant part of thread array
+	if (XBinToFill != -1 && YBinToFill != -1) {
           samplePDFFD_array_private[YBinToFill][XBinToFill] += totalweight;
           samplePDFFD_array_private_w2[YBinToFill][XBinToFill] += totalweight*totalweight;
-		}
-		else{
-		  //std::cout << "Not filled samplePDFFD_array at YBin: " << YBinToFill << " and XBin: " << XBinToFill << std::endl;
-		}
-
-	  }
-	}    
-
-	//End of Calc Weights and fill Array
-	//==================================================
-  // DB Copy contents of 'samplePDFFD_array_private' into 'samplePDFFD_array' which can then be used in GetLikelihood
-	  for (int yBin=0;yBin<nYBins;yBin++) {
-		for (int xBin=0;xBin<nXBins;xBin++) {
+	}
+      }
+    }    
+    
+    //End of Calc Weights and fill Array
+    //==================================================
+    // DB Copy contents of 'samplePDFFD_array_private' into 'samplePDFFD_array' which can then be used in GetLikelihood
+    for (int yBin=0;yBin<nYBins;yBin++) {
+      for (int xBin=0;xBin<nXBins;xBin++) {
 #pragma omp atomic
-		  samplePDFFD_array[yBin][xBin] += samplePDFFD_array_private[yBin][xBin];
+	samplePDFFD_array[yBin][xBin] += samplePDFFD_array_private[yBin][xBin];
```

Comment:
its probably cleaner to have each thread fill its own array that allocated before the parallel region and then after the join, let a single thread add them all up. This isn't a speedy-needing region, but atomics introduce introduce overheads, so its not immediately clear (though maybe you've profiled), that this will be faster when N threads are all fighting to add to the same double, rather than just letting them all write to their own memory and having a single thread do the final sum.